### PR TITLE
Scoped contexts

### PR DESCRIPTION
**Pull request by gkellogg**
Adds support for scoped contexts on properties and types for compaction and expansion.

### CODE REVIEW

File: lib/context.js [hunk 1]
```diff
@@ -979,6 +990,13 @@ function _findContextUrls(input, urls, replace, base) {
             // @context URL found
             urls[_ctx] = false;
           }
+        } else {
+          // look for scoped context
```
Comment (gkellogg) [inline lib/context.js]: @dlongley I think this handles retrieving scoped contexts properly when scanning for contexts, but please consider.

File: lib/compact.js [hunk 2]
```diff
@@ -154,8 +161,16 @@ api.compact = ({
...
-            compactedValue.push(api.compactIri(
-              {activeCtx, iri: expandedValue[vi], relativeTo: {vocab: true}}));
+            const compactedType = api.compactIri(
+              {activeCtx, iri: expandedValue[vi], relativeTo: {vocab: true}})
+
+            // Use any scoped context defined on this value
+              const ctx = _getContextValue(activeCtx, compactedType, '@context');
+              if(ctx) {
+                activeCtx = _processContext({activeCtx, localCtx: ctx, options});
+              }
```
Comment (davidlehn) [inline lib/compact.js]: Too much indent.

File: lib/context.js (hunk 1 above)
Comment (dlongley) [reply lib/context.js]: Heh, I think tests would consider this better than a human. :)

File: lib/context.js [hunk 3]
```diff
@@ -426,6 +426,17 @@ api.createTermDefinition = (activeCtx, localCtx, term, defined) => {
     mapping['@container'] = container;
   }
 
+  // scoped contexts
+  if('@context' in value) {
+    if(activeCtx.processingMode && activeCtx.processingMode.indexOf('json-ld-1.1') !== 0) {
```
Comment (dlongley) [inline lib/context.js]: If we're going to have a decent number of these checks, we should probably make a helper function that takes something like the active context, the processing mode required, and optional error text+details to use if the check fails.

File: lib/expand.js [hunk 4]
```diff
@@ -143,12 +143,31 @@ api.expand = ({
       {activeCtx, localCtx: element['@context'], options});
   }
 
+  let keys = Object.keys(element).sort();
+
+  // look for scoped context on @type
+  for(let ki = 0; ki < keys.length; ++ki) {
```
Comment (dlongley) [inline lib/expand.js]: If you don't need the `ki` index var, use `for...of` and declare your `const key` within it:

```js
for(const key of keys) {
   // ...
```

File: lib/expand.js [hunk 5]
```diff
@@ -143,12 +143,31 @@ api.expand = ({
...
+    const key = keys[ki];
+    let expandedProperty = _expandIri(activeCtx, key, {vocab: true});
+    if(expandedProperty === '@type') {
+
```
Comment (dlongley) [inline lib/expand.js]: Remove empty whitespace line.

File: lib/expand.js [hunk 6]
```diff
@@ -143,12 +143,31 @@ api.expand = ({
       {activeCtx, localCtx: element['@context'], options});
   }
 
+  let keys = Object.keys(element).sort();
+
+  // look for scoped context on @type
+  for(let ki = 0; ki < keys.length; ++ki) {
+    const key = keys[ki];
+    let expandedProperty = _expandIri(activeCtx, key, {vocab: true});
```
Comment (dlongley) [inline lib/expand.js]: Use `const` if `expandedProperty` is not changed later.

File: lib/expand.js [hunk 7]
```diff
@@ -143,12 +143,31 @@ api.expand = ({
...
+    if(expandedProperty === '@type') {
+
+      // set scopped contexts from @type
+      const types = [].concat(element[key]);
+      for(let ii = 0; ii < types.length; ++ii) {
```
Comment (dlongley) [inline lib/expand.js]: Use `for(const type of types)`

File: lib/expand.js [hunk 8]
```diff
@@ -143,12 +143,31 @@ api.expand = ({
...
+    }
+  }
+
   // expand the active property
   const expandedActiveProperty = _expandIri(
     activeCtx, activeProperty, {vocab: true});
 
   let rval = {};
-  let keys = Object.keys(element).sort();
   for(let ki = 0; ki < keys.length; ++ki) {
```
Comment (dlongley) [inline lib/expand.js]: Use `for(const key of keys)`

File: lib/expand.js [hunk 9]
```diff
@@ -313,15 +332,23 @@ api.expand = ({
       continue;
     }
 
-    const container = _getContextValue(activeCtx, key, '@container') || [];
+    // use potential scoped context for key
+    let termCtx = activeCtx;
+    const ctx = _getContextValue(activeCtx, key, '@context');
+    if(ctx) {
+      termCtx = _processContext(
+        {activeCtx, localCtx: ctx, options});
```
Comment (dlongley) [inline lib/expand.js]: Seems like this param would fit on the same line as `_processContext` (i.e. no need to drop it down).

File: lib/context.js (hunk 3 above)
Comment (gkellogg) [reply lib/context.js]: The processingMode check can be replaced with the separate check-td-keys, which will simplify these tests, but a helper function is probably worthwhile.

**Comment by gkellogg** [comment]
I've made the requested changes, and rebased onto check-td-keys. You'll want to be sure that branch is pulled to 0.5.x before this is pulled to 0.6.x.

**Comment by davidlehn** [comment]
Side note: There's a bunch of code here and elsewhere that hasn't changed to for..of or is using let vs const.  I had skipped commenting on that figuring we'd address it in later updates.

**Comment by gkellogg** [comment]
@davidlehn I'd rather not pull this into 0.6.x before #214 is pulled into 0.5, as it depends on the functionality there.

**Pull request merged**